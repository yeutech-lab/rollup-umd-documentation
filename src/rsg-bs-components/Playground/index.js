export { default } from 'rsg-components/Playground/PlaygroundRenderer';
