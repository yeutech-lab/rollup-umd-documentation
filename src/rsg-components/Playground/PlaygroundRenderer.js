import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';

export function PlaygroundRenderer({
  name,
  preview,
  previewProps,
  tabButtons,
  tabBody,
  toolbar,
}) {
  const { ...props } = previewProps;

  return (
    <div>
      <div {...props} data-preview={name}>
        {preview}
      </div>
      <div>
        <div>{tabButtons}</div>
        <div>{toolbar}</div>
      </div>
      <div>{tabBody}</div>
    </div>
  );
}

PlaygroundRenderer.propTypes = {
  name: PropTypes.string.isRequired,
  preview: PropTypes.node.isRequired,
  previewProps: PropTypes.object.isRequired,
  tabButtons: PropTypes.node.isRequired,
  tabBody: PropTypes.node.isRequired,
  toolbar: PropTypes.node.isRequired,
};

export default PlaygroundRenderer;
