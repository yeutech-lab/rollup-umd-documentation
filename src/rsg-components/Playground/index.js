export { default } from '../Playground/Playground';
