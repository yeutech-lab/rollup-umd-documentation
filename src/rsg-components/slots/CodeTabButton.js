import React from 'react';
import PropTypes from 'prop-types';
import TabButton from 'rsg-components/TabButton';

const CodeTabButton = props => <TabButton {...props}>View Code</TabButton>;

CodeTabButton.propTypes = {
	onClick: PropTypes.func.isRequired,
	name: PropTypes.string.isRequired,
	active: PropTypes.bool,
};

export default CodeTabButton;
