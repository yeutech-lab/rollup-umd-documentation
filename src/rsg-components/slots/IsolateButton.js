import React from 'react';
import PropTypes from 'prop-types';
import MdFullscreen from 'react-icons/lib/md/fullscreen';
import MdFullscreenExit from 'react-icons/lib/md/fullscreen-exit';
import ToolbarButton from 'rsg-components/ToolbarButton';
import getUrl from '../../utils/getUrl';

const IsolateButton = ({ name, example, isolated }) =>
	isolated ? (
		<ToolbarButton href={getUrl({ anchor: true, slug: '/' })} title="Show all components">
			<MdFullscreenExit />
		</ToolbarButton>
	) : (
		<ToolbarButton href={getUrl({ name, example, isolated: true })} title="Open isolated">
			<MdFullscreen />
		</ToolbarButton>
	);

IsolateButton.propTypes = {
	name: PropTypes.string.isRequired,
	example: PropTypes.number,
	isolated: PropTypes.bool,
};

export default IsolateButton;
