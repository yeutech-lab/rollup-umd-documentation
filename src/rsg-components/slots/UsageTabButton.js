import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';
import TabButton from '../TabButton'; // eslint-disable-line no-unused-vars

const UsageTabButton = (props) => {
  const component = props.props;
  const showButton = component.props || (component.methods && component.methods.length > 0);
  return showButton ? <TabButton {...props}>Props & methods</TabButton> : null;
};

UsageTabButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  props: PropTypes.shape({
    props: PropTypes.array,
    methods: PropTypes.array,
  }).isRequired,
  active: PropTypes.bool,
};

export default UsageTabButton;
