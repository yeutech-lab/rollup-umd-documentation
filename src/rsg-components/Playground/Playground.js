import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { DisplayModes } from '../../consts';

export default class Playground extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		index: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		settings: PropTypes.object,
	};

	static contextTypes = {
		config: PropTypes.object.isRequired,
		displayMode: PropTypes.string,
	};

	constructor(props, context) {
		super(props, context);
		const { code, settings } = props;
		const { config } = context;
		const showCode = settings.showcode !== undefined ? settings.showcode : config.showCode;

		this.state = {
			code,
			activeTab: showCode ? EXAMPLE_TAB_CODE_EDITOR : undefined,
		};

		this.handleTabChange = this.handleTabChange.bind(this);
		this.handleChange = debounce(this.handleChange.bind(this), config.previewDelay);
	}

	componentWillReceiveProps(nextProps) {
		const { code } = nextProps;
		this.setState({
			code,
		});
	}

	componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	handleChange(code) {
		this.setState({
			code,
		});
	}

	handleTabChange(name) {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	}

	render() {
		const { code, activeTab } = this.state;
		const { evalInContext, index, name, settings } = this.props;
		const { displayMode } = this.context;
		const preview = <Preview code={code} evalInContext={evalInContext} />;
		if (settings.noeditor) {
			return <Para>{preview}</Para>;
		}
		return (
			<PlaygroundRenderer
				name={name}
				preview={preview}
				previewProps={settings.props || {}}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						props={{ code, onChange: this.handleChange }}
					/>
				}
				toolbar={
					<Slot
						name="exampleToolbar"
						props={{ name, isolated: displayMode === DisplayModes.example, example: index }}
					/>
				}
			/>
		);
	}
}
