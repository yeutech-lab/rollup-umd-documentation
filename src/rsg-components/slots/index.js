import Editor from 'rsg-components/Editor';
import Usage from 'rsg-components/Usage';
import IsolateButton from './IsolateButton';
import CodeTabButton from './CodeTabButton';
import UsageTabButton from './UsageTabButton';

export const EXAMPLE_TAB_CODE_EDITOR = 'rsg-code-editor';
export const DOCS_TAB_USAGE = 'rsg-usage';

export default ({ pagePerSection }) => {
	const toolbar = pagePerSection ? [] : [IsolateButton];

	return {
		sectionToolbar: toolbar,
		componentToolbar: toolbar,
		exampleToolbar: toolbar,
		exampleTabButtons: [
			{
				id: EXAMPLE_TAB_CODE_EDITOR,
				render: CodeTabButton,
			},
		],
		exampleTabs: [
			{
				id: EXAMPLE_TAB_CODE_EDITOR,
				render: Editor,
			},
		],
		docsTabButtons: [
			{
				id: DOCS_TAB_USAGE,
				render: UsageTabButton,
			},
		],
		docsTabs: [
			{
				id: DOCS_TAB_USAGE,
				render: Usage,
			},
		],
	};
};
