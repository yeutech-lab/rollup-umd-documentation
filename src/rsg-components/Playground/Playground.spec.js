import React from 'react';
import PropTypes from 'prop-types';
import Playground from './Playground';
import slots, { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { PlaygroundRenderer, styles } from './PlaygroundRenderer';

const evalInContext = a =>
	new Function('require', 'const React = require("react");' + a).bind(null, require); // eslint-disable-line no-new-func
const reactCodeMirrorSelector = '.react-codemirror2';
const code = '<button>OK</button>';
const newCode = '<button>Not OK</button>';
const props = {
	index: 0,
	name: 'name',
	settings: {},
	evalInContext,
	code,
};
const options = {
	context: {
		config: {
			showCode: false,
			highlightTheme: 'base16-light',
		},
		codeRevision: 0,
		slots: slots({}),
	},
	childContextTypes: {
		slots: PropTypes.object.isRequired,
		codeRevision: PropTypes.number.isRequired,
	},
};

it('should render component renderer', () => {
	const actual = shallow(<Playground {...props} />, options);

	expect(actual).toMatchSnapshot();
});

it('should update code via props', () => {
	const actual = shallow(<Playground {...props} />, options);

	expect(actual.state('code')).toEqual(code);

	actual.setProps({
		code: newCode,
	});

	expect(actual.state('code')).toEqual(newCode);
});

it('should update code with debounce', done => {
	const actual = shallow(<Playground {...props} />, {
		context: {
			...options.context,
			config: {
				...options.context.config,
				previewDelay: 1,
			},
		},
	});

	expect(actual.state('code')).toEqual(code);

	actual.instance().handleChange(newCode);

	expect(actual.state('code')).toEqual(code);
	setTimeout(() => {
		expect(actual.state('code')).toEqual(newCode);
		done();
	}, 3);
});

it('should open a code editor', done => {
	const actual = mount(<Playground {...props} />, options);

	expect(actual.find(reactCodeMirrorSelector)).toHaveLength(0);

	actual.find(`button[name="${EXAMPLE_TAB_CODE_EDITOR}"]`).simulate('click');

	setTimeout(() => {
		actual.update();
		expect(actual.find(reactCodeMirrorSelector)).toHaveLength(1);
		done();
	}, 1);
});

it('should not render a code editor if noeditor option passed in example settings', () => {
	const actual = mount(<Playground {...props} settings={{ noeditor: true }} />, options);
	expect(actual.find(`button[name="${EXAMPLE_TAB_CODE_EDITOR}"]`)).toHaveLength(0);
});

it('should open a code editor by default if showcode=true option passed in example settings', () => {
	const actual = mount(<Playground {...props} settings={{ showcode: true }} />, options);
	expect(actual.text()).toMatch('Loading');
});

it('should open a code editor by default if showCode=true option specified in style guide config', () => {
	const actual = mount(<Playground {...props} />, {
		context: {
			...options.context,
			config: {
				...options.context.config,
				showCode: true,
			},
		},
		childContextTypes: options.childContextTypes,
	});
	expect(actual.text()).toMatch('Loading');
});

it('showcode option in example settings should overwrite style guide config option', () => {
	const actual = mount(<Playground {...props} settings={{ showcode: false }} />, {
		context: {
			...options.context,
			config: {
				...options.context.config,
				showCode: true,
			},
		},
		childContextTypes: options.childContextTypes,
	});
	expect(actual.text()).not.toMatch('Loading');
});

it('renderer should render preview', () => {
	const actual = shallow(
		<PlaygroundRenderer
			classes={classes(styles)}
			name="name"
			preview={<div>preview</div>}
			previewProps={{ className: 'pizza', title: 'salami' }}
			tabButtons={<div>tab buttons</div>}
			tabBody={<div>tab body</div>}
			toolbar={<div>toolbar</div>}
		/>
	);

	expect(actual).toMatchSnapshot();
});
