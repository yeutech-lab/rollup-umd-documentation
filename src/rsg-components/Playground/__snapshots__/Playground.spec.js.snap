// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render preview 1`] = `
<div
  className="root"
>
  <div
    className="preview pizza"
    data-preview="name"
    title="salami"
  >
    <div>
      preview
    </div>
  </div>
  <div
    className="controls"
  >
    <div>
      <div>
        tab buttons
      </div>
    </div>
    <div
      className="toolbar"
    >
      <div>
        toolbar
      </div>
    </div>
  </div>
  <div
    className="tab"
  >
    <div>
      tab body
    </div>
  </div>
</div>
`;

exports[`should render component renderer 1`] = `
<Styled(Playground)
  name="name"
  preview={
    <Preview
      code=
<button>
  OK
</button>

      evalInContext={[Function]}
    />
  }
  previewProps={Object {}}
  tabBody={
    <Slot
      active={undefined}
      name="exampleTabs"
      onlyActive={true}
      props={
        Object {
          "code": 
<button>
  OK
</button>
,
          "onChange": [Function],
        }
      }
    />
  }
  tabButtons={
    <Slot
      active={undefined}
      name="exampleTabButtons"
      props={
        Object {
          "onClick": [Function],
        }
      }
    />
  }
  toolbar={
    <Slot
      name="exampleToolbar"
      props={
        Object {
          "example": 0,
          "isolated": false,
          "name": "name",
        }
      }
    />
  }
/>
`;
